Reply to ZIP requests from clients with any network assignment

NJRoadfan opened the pull request:
Some AppleTalk clients do not default to the startup network range and may use a previously connected network number as a hint. Those clients may broadcast a ZIP GetNetInfo packet to determine the network they are connected to. This change allows TashRouter to respond to any client's ZIP requests no matter what network they are reporting from.

NJRoadfan committed:
Allow TashRouter to reply to ZIP requests from clients with any network assignment

Some AppleTalk clients do not default to the startup network range and may use a previously connected network number as a hint. Those clients may broadcast a ZIP GetNetInfo packet to determine the network they are connected to. This change allows TashRouter to respond to any client's ZIP requests no matter what network they are reporting from.

lampmerchant commented:
Hmm.  What happens if we don't respond to the GetNetInfo?  Does the client in question eventually give up and retry with a network number in the startup range?  I'm concerned about sending a reply over the network with an invalid network number...

NJRoadfan commented:
Based on my testing.... it depends on the client. It also doesn't help that Inside AppleTalk is very vague on what a client should actually do. Netatalk ONLY acquires network information from a ZIP GetNetInfo packet, even if it isn't acting as a router. It continually tries ZIP GNI and does NOT glean the information from RTMP packets from Phase 2 networks (only Phase 1/LocalTalk).

After a few tries, it reverts to assuming its on a routerless network and that all network addresses from 0000-FFFE are "valid". It never attempts to reassign the interface in the startup range, which is seemingly "ok" per IA page 5-19. The other vague area in IA is what a router should respond to as a "broadcast". There is a note in there that if someone broadcasts a ZIP GNI that the router should respond with a broadcasted response if it can't determine a source to address the packet to.

I have to do testing with some Mac clients and the Fastpath (particularly how it handles ZIP GNI from nodes outside of its seed range). They may be a bit more forgiving. This is one reason why I'm after that "AppleTalk Phase 2 Protocol Specification" document, it might clear some of this up. I picked socket 6 requests in the conditional as they should be "safe". Although its possible to tighten that up to DDP type 6 and addressed to broadcast "just in case". Other ports should not be sending ZIP GNI and Query across the router, only to the router itself.

I came across this problem when trying to tighten up TashRouter's RTMP validation. Turns out that it will add invalid tuples from non-seed routers outputting stale RTMP data.

NJRoadfan commented:
I did some additional testing.Conclusion: this PR is needed to reduce headaches and it follows how a commercial router handles this. More importantly, Apple expects it!

On a Macintosh (and with netatalk for that matter), the following happens on boot:
-AARP Probe using saved last network address. If this is a fresh install, it picks from the startup range.
-Macintosh sends a ZIP GetNetInfo packet from whatever address it gets from a AARP probe, which may have an old/stale network number.
-If it receives no reply, it continues to send ZIP GNI packets using the same network number, it never switches to the startup range or gleans the current network's data from RTMP.

From what I can tell, both classic networking and Open Transport do NOT glean network data from RTMP broadcasts on extended networks. You can force Open Transport to do so by opening the AppleTalk control panel. Once it sees the new net range, it does the above procedure and connects to the network. Essentially, Apple themselves never uses RTMP data to bootstrap a client on extended networks, only non-extended networks. The only method used for startup configuration is ZIP GNI requests which may not come from the startup range.

I also found out that netatalk has the same problem as TashRouter, although this is likely due to a broadcast handling bug in the Linux kernel driver. The Shiva Fastpath responds to ZIP GNI requests regardless of what source network is in the packet.

lampmerchant commented:
Thanks for doing all this testing and investigation, it's really helpful.

I'm not a huge fan of special-casing ZIP datagrams as a solution, though, it seems like this case is exposing some faulty behavior on the part of `Router.reply`.  I had initially thought to solve it by setting up setting up `Router.route` to return whether or not it was able to route the Datagram and having `Router.reply` respond to a failure to route by sending a reply locally on the receiving Port.  This would deal with the case of an invalid network number or a network number in the startup range, but it notably would _not_ deal with the case of a stale network number that now corresponds to another network somewhere else in the routing table.  In that case, a reply would be sent off to an entirely wrong node, which would be bad.

Trying to think what behavior we really _want_ out of `Router.reply`.  If the source network is in the startup range, we want a reply on the receiving port.  If the source network is a known network, we want the reply to be routed there.  But how do we differentiate that last case from the case of a node coming up and trying its last-known network number?

NJRoadfan commented:
I would go for the following approach:

If a ZIP GNI packet comes into a port, only send a response locally back out that port, do not attempt to route that packet to another network. The conditional to check for the packet would be always coming into an extended network port, always broadcasted, always destination socket 6, always DDP type 6. It would be annoying, but additionally if the above conditional is met, check the first byte of the data packet is 0x05 to confirm it is a ZIP GNI packet type.

Somewhat related, Netatalk filters out "unsolicited" ZIP GetNetInfo replies it receives, which happens all the time due to this stupid Linux multicast bug (literally the kernel is sending netatalk's own responses out the wrong port!). The client (or router) is expected to track when it sends out one of these packets. The reason for this is that GNI is how routers on extended networks are supposed to bootstrap themselves when they are non-seed routers.... but I'll leave that for an issues post later today as that is another bug I ran across. This would require a more robust ZIP GNI implementation, namely sending those packets out and parsing replies.

NJRoadfan committed:
Add more robust packet test

The previous commit didn't allow for cross network ZIP GetNetInfo requests and was a bit too broad. Now we check if a packet is outside of our net range on this port, is being sent to broadcast, and is DDP Type 6 (ZIP).

NJRoadfan commented:
Added a more through check for these packets. I could go further and inspect the first byte to make sure its a ZIP GetNetInfo packet, but I don't think its needed.

lampmerchant commented:
Hmm.  I'm still trying to find a way to avoid special-casing ZIP datagrams, I'm really hoping it's something that _can_ be avoided.

One way to avoid it might be to say that if a datagram comes to the router directly rather than via another router, always answer it directly rather than trying to route it... but to do this, we'd need to check the hop count, and the chart on page 5-26 of IA suggests that datagrams that are routed via one router (for example, A to B on a network that looks like `A---router---B`) will have a hop count of zero.  TashRouter follows that convention (and as long as it does, this approach is out) but I'm wondering, is it the only one?  Like, do AIR and commercial routers increment the hop count when delivering a packet directly over one of their connected networks?

lampmerchant commented:
If I can point to the behavior of other routers (especially AIR) as a precedent for saying that datagrams that come via one router have a hop count of 1, this becomes easy - just change `Router.route` to bump the hop count when routing a packet to a directly-connected destination and change the conditional on line 156 to `if datagram.hop_count == 0 and rx_port.node:`.

NJRoadfan commented:
Keep in mind that ZIP GNI is pretty similar to DHCP. Its a baseline request to bring a node up on a network into a known state. Inside AppleTalk sort of brushes over this, hence why I'm trying to get other documentation that Apple put out on this.

The chart on 5-26 is an ideal scenario. It doesn't accommodate the realities of an actual network and makes no mention of packet handling by the services directly provided by the router (ZIP, NBP). Any node transmitting a ZIP or NBP packet addressed to broadcast on a network would be canned, not processed by a router, if you strictly followed that diagram!

Pretty sure that one node can't broadcast onto another network directly, hence the existence of NBPBrReq and NBPFwdReq. If they can, it basically never happens. Broadcasts should never be forwarded over a router as well and are always local to the network they are made in.

See: https://groups.google.com/g/comp.protocols.appletalk/c/zWhcRdOkWQY/m/0uc782tZtrAJ

> This isn't exactly correct. When a node first comes up and wants to
verify its network info, it will broadcast the GNI request. **Broadcasts
aren't forwarded, though, so the reply will still come from a router on
the local network.**

lampmerchant commented:
> makes no mention of packet handling by the services directly provided by the router (ZIP, NBP).

They're covered by the "Deliver to appropriate socket (Packet addressed to socket in router)" box in the lower right corner.

> Any node transmitting a ZIP or NBP packet addressed to broadcast on a network would be canned

Those are covered by the "Deliver a copy to destination socket in this router" box.

> Broadcasts aren't forwarded

The way I read this is that packets where the DDP destination node is 255 with a network of 0 or the port's network aren't forwarded, which is true, because they've presumably already been broadcast to the network (the LLAP destination is 0xFF or the ELAP destination is 0x090007FFFFFF).  But if a node unicasts a packet to the router with a DDP destination node of 255 on some other network number, the router should route it as it would any other packet bound for that network number, and once it reaches a router directly connected to that network, it should be broadcast.  I don't know of a situation where this would happen either, though, but the flowchart does call for it (broadcast is the destination node in "Send to destination node through routing table entry's port").

Anyway, I'm okay deviating from the flowchart if there's a good reason to do so (especially if other routers do it and _especially_ if AIR/ATIR do it), I just want to know if that's the case where increasing the hop count is concerned.  I can test ATIR myself but I'm not set up to do AIR (I think they're different things though it may be that ATIR is just an older version of AIR?) and I don't have a Shiva/Gatorbox...

NJRoadfan commented:
ATIR = AIR. One thing to keep in mind that if a node can't determine its network number, it gives up and treats 0-0xFFFF as a valid network. There is no requirement to reconfigure itself to the startup range in those cases. Hence, there is always the possibility of a machine popping up with the "wrong network" on a given port on a router.

I think the FastPath increments hop counts, what is a good test case for that?

lampmerchant commented:
Okay, I've been able to confirm that ATIR 2.0 does increment the hop count when delivering a datagram to a node on a network to which it is directly connected!

My test setup is a little convoluted as it was composed of whatever I had lying around...

- Mini vMac running AppleTalk Poke connected via LToUDP to TashRouter
- TashRouter connected via TashTalk to ATIR 2.0
- ATIR 2.0 connected via LocalTalk to:
  - A PIC running PicoATP
  - A PM6100 running AppleTalk Traffic Peek, available 

I tried at first to do this with NetMinder, but for some reason it always reports the hop count as zero.  AppleTalk Peek and Poke are available [here](https://macintoshgarden.org/apps/appletalk-peek-program).  I used Poke to send a datagram from Mini vMac to the PicoATP PIC - the receiver doesn't really matter, it just has to be able to send a CTS frame so the actual payload gets transferred - and looked at it in Peek, and sure enough, the hop count was 2, not 1.

If the FastPath does the same thing (and after this, I'm expecting it will), I think that's justification to change TashRouter as I described above: make it bump the hop count when delivering a datagram to a local net, and use the hop count of a datagram being replied to to determine whether it should be unicast or routed.  What do you think?

NJRoadfan commented:
All DDP hop counts increase by one when transiting across the FastPath. That's how it should work logically with any network router.

Also, looking at how the GNI Query is handled. Both the FastPath and netatalk transmit their GNI Reply via broadcast when they receive a ZIP GNI from a node with a "stale" number, or in the startup range. If the node has a number within the network range of the router's port, it is directly transmitted.

lampmerchant commented:
> Both the FastPath and netatalk transmit their GNI Reply via broadcast when they receive a ZIP GNI from a node with a "stale" number, or in the startup range

Interesting - on both LocalTalk and EtherTalk?

NJRoadfan commented:
ZIP GetNetInfo packets were introduced with Phase 2 and are strictly used on extended networks. They simply aren't needed on non-extended networks as they have only one network number (acquired via RTMP) and one zone. ZIP GNI requests are the only way to determine a network's default zone, hence why they are needed when bootstrapping a non-seed router to a network.

lampmerchant commented:
> ZIP GetNetInfo packets were introduced with Phase 2 and are strictly used on extended networks.

Ah, right, got it.

Okay, I added the same broadcast behavior (as well as bumping the hop count for every datagram the router retransmits): https://github.com/lampmerchant/tashrouter/commit/e587a02e6c38f33734664a19e2bcee8b3fe17aa3

Thanks again for chasing this down.

NJRoadfan commented:
Tested update and appears to be working as intended. Closing.

NJRoadfan commented:
OK, hit a regression. A LocalTalk client is crashing the router with error. This occurs when booting Minivmac:

```
Exception in thread Thread-15:
Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.9/threading.py", line 892, in run
    self._target(*self._args, **self._kwargs)
  File "/home/user1/tashrouter/tashrouter/service/rtmp/responding.py", line 116, in _run
    router.reply(datagram, rx_port, self.RTMP_DDP_TYPE_DATA, response_data)
  File "/home/user1/tashrouter/tashrouter/router/router.py", line 168, in reply
    self.route(Datagram(hop_count=0,
  File "/home/user1/tashrouter/tashrouter/router/router.py", line 149, in route
    entry.port.unicast(datagram.destination_network, datagram.destination_node, datagram.hop())
  File "/home/user1/tashrouter/tashrouter/port/localtalk/__init__.py", line 156, in unicast
    self.send_frame(bytes((node, self.node, self.LLAP_APPLETALK_SHORT_HEADER)) + datagram.as_short_header_bytes())
  File "/home/user1/tashrouter/tashrouter/datagram.py", line 118, in as_short_header_bytes
    raise ValueError('invalid hop count %d, short-header datagrams may not have non-zero hop count' % self.hop_count)
ValueError: invalid hop count 1, short-header datagrams may not have non-zero hop count
```

NJRoadfan commented:
The traffic from MiniVmac was a short DDP "GetMyZone" packet directed to the router. Seems that packet type is triggering something.

```
ALAP: Destination Node = 254
ALAP: Source node = 59
ALAP: LAP protocol type = 1 (Short DDP)
ALAP:
DDP:
DDP: Length = 13
DDP: Destination socket = 6
DDP: Source socket = 247
DDP: DDP protocol type = 3 (ATP)
DDP:
ATP:
ATP: Control = 64
ATP: 01 .. = TRequest
ATP: Bitmap/Sequence Number = 1
0000 0001 = bitmap

ATP: User bytes = 0x07000000
ATP:
ZIP:
ZIP: ZIP command = 7 (GetMyZoneRequest)
ZIP:

ATP:
ATP: Tid = 0x568

ZIP Header
```

Error below:
```
Exception in thread Thread-17:
Traceback (most recent call last):
  File "/usr/lib/python3.9/threading.py", line 954, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.9/threading.py", line 892, in run
    self._target(*self._args, **self._kwargs)
  File "/home/user1/tashrouter/tashrouter/service/zip/responding.py", line 226, in _run
    self._get_my_zone(router, datagram, rx_port)
  File "/home/user1/tashrouter/tashrouter/service/zip/responding.py", line 166, in _get_my_zone
    router.reply(datagram, rx_port, cls.ATP_DDP_TYPE, struct.pack('>BBHBBHB',
  File "/home/user1/tashrouter/tashrouter/router/router.py", line 168, in reply
    self.route(Datagram(hop_count=0,
  File "/home/user1/tashrouter/tashrouter/router/router.py", line 149, in route
    entry.port.unicast(datagram.destination_network, datagram.destination_node, datagram.hop())
  File "/home/user1/tashrouter/tashrouter/port/localtalk/__init__.py", line 156, in unicast
    self.send_frame(bytes((node, self.node, self.LLAP_APPLETALK_SHORT_HEADER)) + datagram.as_short_header_bytes())
  File "/home/user1/tashrouter/tashrouter/datagram.py", line 118, in as_short_header_bytes
    raise ValueError('invalid hop count %d, short-header datagrams may not have non-zero hop count' % self.hop_count)
ValueError: invalid hop count 1, short-header datagrams may not have non-zero hop count
```

NJRoadfan commented:
Commenting out line 117 in datagram.py fixes it.

https://github.com/lampmerchant/tashrouter/blob/e587a02e6c38f33734664a19e2bcee8b3fe17aa3/tashrouter/datagram.py#L117

These were packets directed to the router. Seems that incrementing the hop had a side effect.

lampmerchant commented:
Erp, I forgot we should only be incrementing the hop count on datagrams we _don't_ originate.  Pushed a fix for that here: https://github.com/lampmerchant/tashrouter/commit/4ae50053cd2908ed9c5a94db6218d33d3bf33755

NJRoadfan commented:
I brought TashRouter up on a macvtap interface and now have it directly connected to the FastPath.... and there is a slight problem. Directly addressed ZIP Queries and ZIP GetNetworkInfo packets aren't working for some reason. I see both TashRouter and the Fastpath trying to update their zone tables issuing direct ZIP Queries to each other (not broadcast, those seem to work)..... and neither one is responding to each other. This is after the FastPath acquires a network and default zone, so it is on the same network, not the startup range or an invalid number. I'm beginning to wonder if this is some macvtap weirdness as I'm running Linux in a VirtualBox VM.

This is working fine with atalkd from Netatalk via a standard tap interface, so something is up.

NJRoadfan commented:
OK, its macvtap weirdness. Only broadcast packets are being seen across the link... weird.

lampmerchant commented:
> OK, its macvtap weirdness. Only broadcast packets are being seen across the link... weird.

Did you have promiscuous mode set on in the macvtap?  My memory's a little fuzzy on the subject, but I seem to remember broadcast packets had something to do with why it was necessary...
